### Conclusions  - web_page_dataset_phishing:

The dataset is available in https://www.kaggle.com/datasets/shashwatwork/web-page-phishing-detection-dataset


- Explain what conclusions you would draw from this analysis: are the data what you expect?  Are the data likely to be usable?  If the data are not useable, find some new data!

  This is a very good dataset and is highly usable for doing the phishing related EDA and we can build models to identify fraudulent /phishing websites based on the features provided

- Do you see any outliers?  (Data points that are far from the rest of the data).

  Based on the Box Plot and Violin plot we can clearly see the outliers in multiple features like nb_slash , nb_hyphen , nb_dots and many other

- Are any data items highly correlated with each other, suggesting that they are redundant?

  There are multiple pairs of features where correlation values were very high like longest_words_raw & longest_word_path pair (0.92 correlation factor) , nb_and & nb_eq pair (0.86). There are many more and that gives us the opportunity to do feature reduction

- For the line plots, do you see a trend or pattern over time?  Does this suggest that the data are changing over time (drifting) in such a way as to invalidate comparisons?

  This Phishing dataset did not have any date in the features so we will not be able to comment

- Can you think of any confounding variables?  (Third variables that could explain any correlations between other variables.  These third variables may or may not be reported in the dataset.)

  No we could not find any confounding variable right now

In [74]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import os


In [75]:
#import dataset file "web_page_dataset_phishing.csv" from folder./Phishing_data
path = './dataset/'
os.listdir(path)

#first open the two datasets related to phishing_website and phishing_email

df_phishing_website_data=pd.read_csv(path+'web_page_dataset_phishing.csv')

In [76]:
df_phishing_website_data.shape

(11430, 89)

In [77]:
df_phishing_website_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11430 entries, 0 to 11429
Data columns (total 89 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   url                         11430 non-null  object 
 1   length_url                  11430 non-null  int64  
 2   length_hostname             11430 non-null  int64  
 3   ip                          11430 non-null  int64  
 4   nb_dots                     11430 non-null  int64  
 5   nb_hyphens                  11430 non-null  int64  
 6   nb_at                       11430 non-null  int64  
 7   nb_qm                       11430 non-null  int64  
 8   nb_and                      11430 non-null  int64  
 9   nb_or                       11430 non-null  int64  
 10  nb_eq                       11430 non-null  int64  
 11  nb_underscore               11430 non-null  int64  
 12  nb_tilde                    11430 non-null  int64  
 13  nb_percent                  114

#### Of the 88 columns we have two columns which are neither int or float ie status and url

In [78]:
df_phishing_website_data.describe()

length_url  length_hostname            ip       nb_dots  \
count  11430.000000     11430.000000  11430.000000  11430.000000   
mean      61.126684        21.090289      0.150569      2.480752   
std       55.297318        10.777171      0.357644      1.369686   
min       12.000000         4.000000      0.000000      1.000000   
25%       33.000000        15.000000      0.000000      2.000000   
50%       47.000000        19.000000      0.000000      2.000000   
75%       71.000000        24.000000      0.000000      3.000000   
max     1641.000000       214.000000      1.000000     24.000000   

         nb_hyphens         nb_at         nb_qm        nb_and    nb_or  \
count  11430.000000  11430.000000  11430.000000  11430.000000  11430.0   
mean       0.997550      0.022222      0.141207      0.162292      0.0   
std        2.087087      0.155500      0.364456      0.821337      0.0   
min        0.000000      0.000000      0.000000      0.000000      0.0   
25%        0.000000      0.000000      0.000000      0.000000      0.0   
50%        0.000000      0.000000      0.000000      0.000000      0.0   
75%        1.000000      0.000000      0.000000      0.000000      0.0   
max       43.000000      4.000000      3.000000     19.000000      0.0   

              nb_eq  ...   empty_title  domain_in_title  \
count  11430.000000  ...  11430.000000     11430.000000   
mean       0.293176  ...      0.124759         0.775853   
std        0.998317  ...      0.330460         0.417038   
min        0.000000  ...      0.000000         0.000000   
25%        0.000000  ...      0.000000         1.000000   
50%        0.000000  ...      0.000000         1.000000   
75%        0.000000  ...      0.000000         1.000000   
max       19.000000  ...      1.000000         1.000000   

       domain_with_copyright  whois_registered_domain  \
count           11430.000000             11430.000000   
mean                0.439545                 0.072878   
std                 0.496353                 0.259948   
min                 0.000000                 0.000000   
25%                 0.000000                 0.000000   
50%                 0.000000                 0.000000   
75%                 1.000000                 0.000000   
max                 1.000000                 1.000000   

       domain_registration_length    domain_age   web_traffic    dns_record  \
count                11430.000000  11430.000000  1.143000e+04  11430.000000   
mean                   492.532196   4062.543745  8.567566e+05      0.020122   
std                    814.769415   3107.784600  1.995606e+06      0.140425   
min                     -1.000000    -12.000000  0.000000e+00      0.000000   
25%                     84.000000    972.250000  0.000000e+00      0.000000   
50%                    242.000000   3993.000000  1.651000e+03      0.000000   
75%                    449.000000   7026.750000  3.738455e+05      0.000000   
max                  29829.000000  12874.000000  1.076799e+07      1.000000   

       google_index     page_rank  
count  11430.000000  11430.000000  
mean       0.533946      3.185739  
std        0.498868      2.536955  
min        0.000000      0.000000  
25%        0.000000      1.000000  
50%        1.000000      3.000000  
75%        1.000000      5.000000  
max        1.000000     10.000000  

[8 rows x 87 columns]

In [79]:
#lets us check for null values
df_phishing_website_data.isnull().sum()

url                0
length_url         0
length_hostname    0
ip                 0
nb_dots            0
                  ..
web_traffic        0
dns_record         0
google_index       0
page_rank          0
status             0
Length: 89, dtype: int64

In [80]:
# let us check if there are any rows with null values
df_phishing_website_data[df_phishing_website_data.isnull().any(axis=1)]


Empty DataFrame
Columns: [url, length_url, length_hostname, ip, nb_dots, nb_hyphens, nb_at, nb_qm, nb_and, nb_or, nb_eq, nb_underscore, nb_tilde, nb_percent, nb_slash, nb_star, nb_colon, nb_comma, nb_semicolumn, nb_dollar, nb_space, nb_www, nb_com, nb_dslash, http_in_path, https_token, ratio_digits_url, ratio_digits_host, punycode, port, tld_in_path, tld_in_subdomain, abnormal_subdomain, nb_subdomains, prefix_suffix, random_domain, shortening_service, path_extension, nb_redirection, nb_external_redirection, length_words_raw, char_repeat, shortest_words_raw, shortest_word_host, shortest_word_path, longest_words_raw, longest_word_host, longest_word_path, avg_words_raw, avg_word_host, avg_word_path, phish_hints, domain_in_brand, brand_in_subdomain, brand_in_path, suspecious_tld, statistical_report, nb_hyperlinks, ratio_intHyperlinks, ratio_extHyperlinks, ratio_nullHyperlinks, nb_extCSS, ratio_intRedirection, ratio_extRedirection, ratio_intErrors, ratio_extErrors, login_form, external_favicon, links_in_tags, submit_email, ratio_intMedia, ratio_extMedia, sfh, iframe, popup_window, safe_anchor, onmouseover, right_clic, empty_title, domain_in_title, domain_with_copyright, whois_registered_domain, domain_registration_length, domain_age, web_traffic, dns_record, google_index, page_rank, status]
Index: []

[0 rows x 89 columns]

#### There are no rows with null values

In [81]:
#check if there are duplicate rows
df_phishing_website_data.duplicated().sum()

np.int64(0)

#### There are no duplicate rows

In [82]:
#lets check how many unique values are there in status
df_phishing_website_data['status'].unique()


array(['legitimate', 'phishing'], dtype=object)

In [83]:
df_phishing_website_data['status'].value_counts()

status
legitimate    5715
phishing      5715
Name: count, dtype: int64

### The dataset is balanced dataset with equal number of legitimate and phishing site information

In [84]:
#let us change the dtype of status to int where 0 is 'legitimate' and 1 is 'phishing'
df_phishing_website_data['status'] = df_phishing_website_data['status'].apply(lambda x: 1 if x.split()[0] == 'phishing' else 0)
df_phishing_website_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11430 entries, 0 to 11429
Data columns (total 89 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   url                         11430 non-null  object 
 1   length_url                  11430 non-null  int64  
 2   length_hostname             11430 non-null  int64  
 3   ip                          11430 non-null  int64  
 4   nb_dots                     11430 non-null  int64  
 5   nb_hyphens                  11430 non-null  int64  
 6   nb_at                       11430 non-null  int64  
 7   nb_qm                       11430 non-null  int64  
 8   nb_and                      11430 non-null  int64  
 9   nb_or                       11430 non-null  int64  
 10  nb_eq                       11430 non-null  int64  
 11  nb_underscore               11430 non-null  int64  
 12  nb_tilde                    11430 non-null  int64  
 13  nb_percent                  114

In [85]:
#lets check count of unique values in url
df_phishing_website_data['url'].unique().shape

(11429,)

In [86]:
# Create a random sample of 1000 rows while maintaining class balance
def create_balanced_sample(df, sample_size=500, random_state=42):
    # calculate the number of samples needed from each class
    samples_per_class = sample_size // 2

    # split the data by class
    spam = df[df['status'] == 1]
    non_spam = df[df['status'] == 0]

    # sample equally from each class
    spam_sample = spam.sample(n=samples_per_class, random_state=random_state)
    non_spam_sample = non_spam.sample(n=samples_per_class, random_state=random_state)

    # combine the samples
    balanced_sample = pd.concat([spam_sample, non_spam_sample])

    # chuffle the combined sample
    balanced_sample = balanced_sample.sample(frac=1, random_state=random_state).reset_index(drop=True)

    return balanced_sample

# create the balanced sample
sampled_df = create_balanced_sample(df_phishing_website_data, sample_size= 10000, random_state=39)

# verify the sample
print("Original dataset shape:", df_phishing_website_data.shape)
print("Sampled dataset shape:", sampled_df.shape)
print("\nClass distribution in sample:")
print(sampled_df['status'].value_counts())

# run the analysis on the sampled data
#analyze_phishing_data(sampled_df)

Original dataset shape: (11430, 89)
Sampled dataset shape: (10000, 89)

Class distribution in sample:
status
0    5000
1    5000
Name: count, dtype: int64


### we will not use the URL info as it looks to be unique URL and we see total 11429 unique urls


In [87]:
# lets take the sampled df and drop the url column
sampled_df=sampled_df.drop(columns=['url'])

In [88]:
def remove_constant_columns(df):
    return df.loc[:, (df != df.iloc[0]).any()]

def perform_univariate_analysis(df):
    # calculate number of rows and columns for subplots
    n_features = len(df.columns)
    n_cols = 2
    n_rows = (n_features + n_cols - 1) // n_cols

    # create figure with subplots
    fig, axes = plt.subplots(n_rows, n_cols, figsize=(15, 6*n_rows))
    fig.suptitle('Univariate Analysis of Features', fontsize=16, y=1.02)

    # flatten axes array for easier iteration
    axes = axes.flatten()

    # create histograms and KDE plots for each feature
    for idx, column in enumerate(df.columns):
        # Histogram with KDE
        sns.histplot(data=df, x=column, hue='status', multiple="stack", kde=True, ax=axes[idx])
        axes[idx].set_title(f'Distribution of {column}')
        axes[idx].set_xlabel(column)
        axes[idx].set_ylabel('Count')

    # remove empty subplots if any
    for idx in range(len(df.columns), len(axes)):
        fig.delaxes(axes[idx])

    plt.tight_layout()
    plt.show()

In [89]:
# Method to do bivariate analysis

def perform_bivariate_analysis(df):
      # create correlation Matrix
    plt.figure(figsize=(12, 8))
    correlation_matrix = df.corr()
    sns.heatmap(correlation_matrix,
                annot=True,
                cmap='magma',
                center=0,
                fmt='.2f',
                square=True)
    plt.title('Correlation Matrix of Features')
    plt.tight_layout()
    plt.show()

    significant_correlations = create_correlation_df(correlation_matrix, threshold=0.5)
    print("Significant Correlation values : ", significant_correlations)

    # pairplot with KDE
    #print("generating pairplot ...")
    #sns.pairplot(df, hue='status', diag_kind='kde')
    #plt.show()

    # violin plots
    features = [col for col in df.columns if col != 'status']
    n_features = len(features)
    n_cols = 2
    n_rows = (n_features + n_cols - 1) // n_cols

    fig, axes = plt.subplots(n_rows, n_cols, figsize=(15, 6*n_rows))
    fig.suptitle('Violin Plots by Status', fontsize=16, y=1.02)
    axes = axes.flatten()

    for idx, feature in enumerate(features):
        sns.violinplot(data=df, x='status', y=feature, ax=axes[idx])
        axes[idx].set_title(f'{feature} by Status')

    # remove empty subplots
    for idx in range(len(features), len(axes)):
        fig.delaxes(axes[idx])

    plt.tight_layout()
    plt.show()

    # generate box plots
    fig, axes = plt.subplots(n_rows, n_cols, figsize=(15, 6*n_rows))
    fig.suptitle('Box Plots by Status', fontsize=16, y=1.02)
    axes = axes.flatten()

    for idx, feature in enumerate(features):
        sns.boxplot(data=df, x='status', y=feature, ax=axes[idx])
        axes[idx].set_title(f'{feature} by Status')

    # Remove empty subplots
    for idx in range(len(features), len(axes)):
        fig.delaxes(axes[idx])

    plt.tight_layout()
    plt.show()

In [90]:
def perform_kde_analysis(df):
    features = [col for col in df.columns if col != 'status']
    n_features = len(features)
    n_cols = 2
    n_rows = (n_features + n_cols - 1) // n_cols

    fig, axes = plt.subplots(n_rows, n_cols, figsize=(15, 6*n_rows))
    fig.suptitle('KDE Plots by Feature', fontsize=16, y=1.02)
    axes = axes.flatten()

    for idx, feature in enumerate(features):
        sns.kdeplot(data=df, x=feature, hue='status', ax=axes[idx])
        axes[idx].set_title(f'KDE Plot of {feature}')

    # Remove empty subplots
    for idx in range(len(features), len(axes)):
        fig.delaxes(axes[idx])

    plt.tight_layout()
    plt.show()

In [91]:
def analyze_phishing_data(df):
	#execute it in the below sequence
	print("Performing Univariate Analysis...")
	perform_univariate_analysis(df)
	print("\nPerforming Bivariate Analysis...")
	perform_bivariate_analysis(df)
	print("\nPerforming KDE Analysis...")
	perform_kde_analysis(df)

In [92]:
def analyze_phishing_web_Univariate_data(df):
	print("Performing Univariate Analysis...")
	perform_univariate_analysis(df)


In [93]:
def analyze_phishing_web_Bivariate_data(df):
	print("\nPerforming Bivariate Analysis...")
	perform_bivariate_analysis(df)
	print("\nPerforming KDE Analysis...")
	perform_kde_analysis(df)

In [94]:
def create_correlation_df(correlation_matrix, threshold=0.5):

    #  creates a DataFrame of significant correlations.

    correlations = []

    for i in range(len(correlation_matrix.columns)):
        for j in range(i + 1, len(correlation_matrix.columns)):
            corr_value = correlation_matrix.iloc[i, j]
            if abs(corr_value) >= threshold:
                correlations.append({
                    'Feature 1': correlation_matrix.columns[i],
                    'Feature 2': correlation_matrix.columns[j],
                    'Correlation': corr_value
                })

    corr_df = pd.DataFrame(correlations)
    return corr_df.sort_values('Correlation', key=abs, ascending=False)


In [95]:
df_phishing_web_cleaned = remove_constant_columns(sampled_df)

In [96]:
#analyze_phishing_web_Univariate_data(df_phishing_web_cleaned)

In [97]:
# analyze_phishing_web_Bivariate_data(df_phishing_web_cleaned)

In [98]:
#df_phishing_web_cleaned = remove_constant_columns(sampled_df)
#analyze_phishing_data(df_phishing_web_cleaned)